feat(tooling): add-logic-to-generate-changelogs

Shreyas281299 opened the pull request:
<!--
Hey there,\
Thank you for taking the time to improve our code! 🙂\
Please let us know why this change is necessary and what testing you have performed. \
This ensures our reviewers understand the impact of your change. \

**IMPORTANT**
FAILING TO FILL OUT THIS TEMPLATE WILL RESULT IN REJECTION OF YOUR PULL REQUEST
This is for compliance purposes with FedRAMP program.
-->

# COMPLETES #[SPARK-532257](https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-532257)

## This pull request addresses
Added a new command in the package-tools to create changelogs.
The command is run on deploy after npm publish has already happened.

## by making the following changes
When changelog mehtod is called a changelog file is created or the existing changelog file is updated in the documentation branch.

<!-- You may include screenshots -->

### Change Type

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [x] Tooling change
- [ ] Internal code refactor

## The following scenarios where tested

Updated pull-request.yml to run the changelog command in my fork. 

- https://github.com/Shreyas281299/webex-js-sdk/actions/runs/9807909431/job/27082945372 - Here is the github action that created the following commit
-https://github.com/Shreyas281299/webex-js-sdk/tree/documentation - Here is the commit that added the changelog. 

### I certified that

- [x] I have read and followed [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request)
- [x] I discussed changes with code owners prior to submitting this pull request

- [x] I have not skipped any automated checks
- [x] All existing and new tests passed
- [x] I have updated the documentation accordingly

---

Make sure to have followed the [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request) before submitting.

mkesavan13 commented on packages/tools/package/src/commands/changelog/changelog.constants.ts:
```diff
@@ -0,0 +1,35 @@
+      type: 'string...',
+    },
+    {
+      description: 'Tag to update this local package with on npm',
```
Could we please reword this?

Shreyas281299 replied:
Updated.

mkesavan13 commented on packages/tools/package/src/commands/changelog/changelog.types.ts:
```diff
@@ -0,0 +1,46 @@
+  [packageName: string]: {
+    [version: string]: {
+      commits: CommitDetails;
+      alongWith: AlongWithData;
```
We also need the `published_date` type added. It should be a Unix timestamp

Shreyas281299 replied:
Added

mkesavan13 commented on packages/tools/package/test/module/changelog/changelog.fixtures.js:
```diff
@@ -0,0 +1,39 @@
+const packagesData = {
+  webex: {
+    data: {
+      location: '/Users/shreysh2/Desktop/Repo/webex-js-sdk/packages/webex',
```
Why is your system details here? Should we consider replacing it with a generic one?

Shreyas281299 replied:
Removed.

mkesavan13 commented on packages/tools/package/test/module/changelog/changelog.fixtures.js:
```diff
@@ -0,0 +1,39 @@
+  },
+  '@webex/package-tools': {
+    data: {
+      location: '/Users/shreysh2/Desktop/Repo/webex-js-sdk/packages/tools/package',
```
Same here

Shreyas281299 replied:
Removed

rarajes2 commented on packages/tools/package/src/commands/changelog/changelog.constants.ts:
```diff
@@ -0,0 +1,35 @@
+    {
+      description: 'Packages to be added in changelog',
+      name: 'packages',
+      type: 'string...',
```
Is it a typo? `string...`

rarajes2 commented on packages/tools/package/src/commands/changelog/changelog.ts:
```diff
@@ -0,0 +1,54 @@
+
+    const tag = options.tag?.split('/').pop();
+    return Yarn.list()
+      .then((packageDetails) => packageDetails.map(
```
Since we have already declared the function as async, we can use `async-await` syntax

Shreyas281299 replied:
Removed async-await. Trying to keep all the tooling consistent with the .then approach

rarajes2 replied:
I don't think we need to maintain consistency with thenable syntax. We should use async-await everywhere now, it's more readable.

rarajes2 commented on packages/tools/package/src/commands/changelog/changelog.ts:
```diff
@@ -0,0 +1,54 @@
+import { createOrUpdateChangelog } from './changelog.utils';
+
+/**
+ * @public
```
We can add description here

Shreyas281299 replied:
I've followed the style with which comments are written in the tooling. The description is written inside the constant so here line 17-20 has the description

rarajes2 commented on packages/tools/package/src/commands/changelog/changelog.types.ts:
```diff
@@ -0,0 +1,46 @@
+  [key: string]: string;
+}
+
+export interface AlongWithData {
```
Q. What does this `AlongWithData` means?

Shreyas281299 replied:
Details about the alongwithdata is given in line 33-36

rarajes2 commented on packages/tools/package/src/commands/changelog/changelog.utils.ts:
```diff
@@ -0,0 +1,118 @@
+ * @param scriptPath - Path to the script which we want to run.
+ * @returns - Promise that resolves once the script is run and returns the output.
+ */
+export function runShellScript(scriptPath: string) {
```
This seems generic function. We can check if this is defined somewhere else as well

Shreyas281299 replied:
Yes, there is an executor class that is used to run command. Using that class instead on the shell script.

rarajes2 commented on packages/tools/package/src/commands/changelog/changelog.utils.ts:
```diff
@@ -0,0 +1,118 @@
+/**
+ * Function to get and formate alongWith object
+ * alongWith object has the information about packages that are modified `along with`
+ * a package
```
Description could be a bit more structured

Shreyas281299 replied:
Ive tried to make the description a bit more clear.

sreenara commented on .github/workflows/deploy.yml:
```diff
@@ -254,7 +254,12 @@ jobs:
           git config user.name "${GIT_AUTHOR_NAME}"
 
-      - name: Publish Docs Folder
+      - name: Updated changelog
```
```suggestion
      - name: Update changelog
```

Shreyas281299 replied:
Changed

sreenara commented on .github/workflows/pull-request.yml:
```diff
@@ -168,7 +168,6 @@ jobs:
           path: '**/dist'
           key: dist-${{ env.rid }}
 
-
```
This change isn't required right? Can we revert it

Shreyas281299 replied:
Removed

sreenara commented on packages/tools/package/src/commands/changelog/changelog.utils.ts:
```diff
@@ -0,0 +1,118 @@
+}
+
+/**
+ * Function to get and formate alongWith object
```
Typo in the word format

Shreyas281299 replied:
Fixed

sreenara commented on packages/tools/package/package.json:
```diff
@@ -30,6 +30,7 @@
     "start": "node ./index.js",
     "test": "yarn test:style && yarn test:syntax && yarn test:integration && yarn test:coverage",
     "test:coverage": "yarn test:integration --coverage --reporters=\"jest-silent-reporter\"",
+    "test:unit": "jest",
```
Why was this needed? UTs were running for the other commands in this package before right?

Shreyas281299 replied:
We run the `test:unit` command to run unit tests for all the packages but here we didn't have the command. Although nothing is broken because test:integration is running the unit test as you can see in the command below.

sreenara replied:
@Shreyas281299 let's remove the test:integration command then. It's not needed.

sreenara reviewed (changes requested):
Please see the comments. Also, can we add these tests?
1. Where the changelog is already present and a new changelog is appended?
2. We move to a new version such as 3.4.0 or 4.0.0 and a new file is created?

Shreyas281299 commented:
> Please see the comments. Also, can we add these tests?
> 
>     1. Where the changelog is already present and a new changelog is appended?
> 
>     2. We move to a new version such as 3.4.0 or 4.0.0 and a new file is created?

Added test. These tests were missing I have added a new file with these test - changelog.utils.test

sreenara commented on packages/tools/package/test/module/changelog/changelog.utils.test.js:
```diff
@@ -0,0 +1,66 @@
+const changelogUtils = require('@webex/package-tools/dist/module/commands/changelog/changelog.utils');
+const changelogExecutor = require('@webex/package-tools/dist/module/commands/changelog/changelog.executor');
```
Why do we have to require this way with a complete path? Does relative path not work? Is there some error?

Shreyas281299 replied:
Yes. Im unable to import them relatively. I have used this also to maintain coding convention in packag-tools. This is exactly how utils have been tested in this package.

sreenara commented on packages/tools/package/test/module/changelog/changelog.utils.test.js:
```diff
@@ -0,0 +1,66 @@
+      await changelogUtils.createOrUpdateChangelog(fixtures.packagesData, 'mock_commit_id');
+      expect(fs.readFileSync).not.toHaveBeenCalledWith();
+      expect(fs.mkdirSync).toHaveBeenCalledWith('./docs/changelog', {
+        recursive: true,
```
What's the filename of the new file? Are we not going to check that?

Shreyas281299 replied:
Actually mkdirSync was used as a fallback to create the changelog folder if its not present. And it is fs.writeFileSync that creates or edits the files. So I have removed this method and the test.

sreenara commented on packages/tools/package/test/module/changelog/changelog.utils.test.js:
```diff
@@ -0,0 +1,66 @@
+      });
+    });
+
+    it('creates correct changelog data and writes to the respective changelog file', async () => {
```
I think we could reword this test. If I understand correctly, this test creates a new changelog file and fills it with the right content, right?

Shreyas281299 replied:
Yes. It creates the data, and either populates the existing file or creates a new file and puts the data their.

Shreyas281299 replied:
I have updated the description. Please see

sreenara commented on packages/tools/package/test/module/changelog/changelog.utils.test.js:
```diff
@@ -0,0 +1,66 @@
+        .split('-')[0]
+        .replace(/\./g, '_');
+      expect(fs.writeFileSync).toHaveBeenCalledTimes(2);
+      expect(fs.writeFileSync.mock.calls).toEqual([
```
Can we have a comment that explains these 2 calls we are expecting?

sreenara reviewed (changes requested):
Please see the comments

sreenara commented on packages/tools/package/test/module/changelog/changelog.utils.test.js:
```diff
@@ -0,0 +1,60 @@
+        .split('-')[0]
+        .replace(/\./g, '_');
+      expect(fs.writeFileSync).toHaveBeenCalledTimes(2);
+      // writeFileSync will be called for each package
+      // Here, according to the fixtures first it will be called for webex
+      // and second, it will be called for @webex/package-tools
```
```suggestion
      // writeFileSync will be called for each package
      // Here, according to the fixtures first it will be called for webex which is at version 3.3.1
      // and second, it will be called for @webex/package-tools at version 0.0.0
```

Shreyas281299 replied:
Updated

sreenara reviewed (approved):
Approved with a nitpick in the comment

sreenara reviewed (changes requested):
Can we check why the UT is failing (1 failure)?
Coverage is also failing.

Shreyas281299 commented:
> Can we check why the UT is failing (1 failure)? Coverage is also failing.

Coverage was failling because in test:coverage we were calling test:integration and since that command has been removed it was failling I have replaced test:integration with test:unit

UTs were failling because i didnt handle the existing changelog data in uts properly. Fixed that as well.

rarajes2 commented on .github/workflows/deploy.yml:
```diff
@@ -254,7 +254,12 @@ jobs:
+          git checkout documentation docs/changelog/
+          yarn package-tools changelog --packages ${{ needs.generate-package-matrix.outputs.node-recursive }} --tag ${GITHUB_REF##*/} --commit ${{ github.event.before }}
+
+      - name: Update Changelog and Publish Docs Folder
```
Are we not following any convention for the name description here? Somewhere we are using PascalCase and somewhere it's normal sentence.

Shreyas281299 replied:
I think I dont think we are following any convention here. We can create a task to fix all these.

rarajes2 commented on packages/tools/package/src/commands/changelog/changelog.utils.ts:
```diff
@@ -0,0 +1,79 @@
+  Object.keys(packages).forEach(async (index: any) => {
+    const pkgName = packages[index].name;
+    // const {version} = packages[index];
+    // console.log(packages[index], version);
```
We can remove all the commented lines

Shreyas281299 replied:
Removed

rarajes2 commented on packages/tools/package/src/commands/changelog/changelog.utils.ts:
```diff
@@ -0,0 +1,79 @@
+    const fileName = version.split('-')[0].replace(/\./g, '_');
+    // Constructing the changelog file name
+    const changelogFilePath = `./docs/changelog/v${fileName}.json`;
+    // Prepare the changelog entry
```
Let's leave a blank line before (wherever we add a comment)

Shreyas281299 replied:
Added.

sreenara commented:
> I don't think we need to maintain consistency with thenable syntax. We should use async-await everywhere now, it's more readable.

@rarajes2 let's stick to the conventions here since we have 4 other commands following the same design. It's easy for a person to understand the structure once they've got a hang of it.

rarajes2 reviewed (approved):
LGTM

mkesavan13 reviewed (approved):
Changes look good to me. No red flags. Let's merge it
